### PR TITLE
`cutlass` integration + `segment_matmul` implementation

### PR DESCRIPTION
**Pull request by rusty1s**

### CODE REVIEW

**Comment by rusty1s** [comment]
@pyg-team/nvidia-team This PR is now ready to review.

**Review by teju85 — COMMENTED** [review]
This is really a great example showing cutlass integration. Nice job @rusty1s ! Do you have an example where the `pyg_lib.segment.grouped_matmul` is actually getting called in a training script?

**Review by puririshi98 — APPROVED** [review]
LGTM!

File: pyg_lib/csrc/segment/cuda/matmul_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,157 @@
+#include <ATen/ATen.h>
```
Comment (yaoyaowd) [inline pyg_lib/csrc/segment/cuda/matmul_kernel.cu:1]: I'd suggest use `""` instead `<>` for non-system library.
```
#include "ATen/ATen.h"
#include "cutlass/gemm/device/gemm_grouped.h"
```
We noticed huge compile pain when building third party package with `include <>` in other repos.

File: pyg_lib/csrc/segment/matmul.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#include "matmul.h"
+
+#include <ATen/core/dispatch/Dispatcher.h>
```
Comment (yaoyaowd) [inline pyg_lib/csrc/segment/matmul.cpp:3]: ditto.

**Review by yaoyaowd — APPROVED** [review]

File: pyg_lib/csrc/segment/cuda/matmul_kernel.cu (hunk 1 above)
Comment (rusty1s) [reply pyg_lib/csrc/segment/cuda/matmul_kernel.cu:1]: `ATen` is expect to be a system library.

**Comment by rusty1s** [comment]
Thanks @teju85. I will work on backward implementation and PyG integration next. Can share an example by then!

File: pyg_lib/csrc/segment/cuda/matmul_kernel.cu (hunk 1 above)
Comment (yaoyaowd) [reply pyg_lib/csrc/segment/cuda/matmul_kernel.cu:1]: ATen is from pytorch, https://github.com/pytorch/pytorch/tree/master/aten/src

File: pyg_lib/csrc/segment/cuda/matmul_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  ptr_C.reset(num_matrices);
+  ptr_C.copy_from_host(ptr_C_host.data());
+
+  std::vector<cutlass::gemm::GemmCoord> all_problems(num_matrices);
+  std::vector<int64_t> ld_A_host(num_matrices);
+  std::vector<int64_t> ld_B_host(num_matrices);
+  std::vector<int64_t> ld_C_host(num_matrices);
+
+  for (size_t i = 0; i < num_matrices; ++i) {
+    auto m = input[i].size(0), k = input[i].size(1), n = out[i].size(1);
```
Comment (ZenoTan) [inline pyg_lib/csrc/segment/cuda/matmul_kernel.cu:77]: Could we add some checks about dim here?

File: pyg_lib/csrc/segment/cuda/matmul_kernel.cu [hunk 4]
```diff
@@ -0,0 +1,157 @@
+#include <ATen/ATen.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <torch/library.h>
+
+#include <cutlass/gemm/device/gemm_grouped.h>
+#include <cutlass/gemm/kernel/default_gemm_grouped.h>
+#include <cutlass/util/host_tensor.h>
+
+namespace pyg {
+namespace segment {
```
Comment (ZenoTan) [inline pyg_lib/csrc/segment/cuda/matmul_kernel.cu:10]: My feeling is that `pyg::segment` might better be something like `pyg::ops::segment`. Want to hear about how you would organize future kernels

**Comment by hwu36** [comment]
Haicheng from nvidia cutlass.   LGTM.  Thank you.  BTW, we are improving group gemm now.

**Comment by rusty1s** [comment]
@hwu36 Thanks! Please ping me if you make any improvements :)

**Comment by hwu36** [comment]
> @hwu36 Thanks! Please ping me if you make any improvements :)

@jackkosaian just fixed occupancy calculation in https://github.com/NVIDIA/cutlass/pull/532 .  This number is used to calculated the number of threadblocks to launch group gemm.  I know you hard coded this number now so you are not affected.

@jackkosaian is going to further improve group gemm in the summer.
